rcl_action/include/rcl_action/action_server.h

```diff
@@ -0,0 +1,727 @@
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] action_server handle to the action server that will take the cancel request
+ * \param[out] ros_cancel_request a preallocated ROS cancel request where the request
+ *   message is copied
+ * \return `RCL_RET_OK` if the response was sent successfully, or
+ * \return `RCL_RET_INVALID_ARGUMENT` if any arguments are invalid, or
+ * \return `RCL_RET_ACTION_SERVER_INVALID` if the action server is invalid, or
+ * \return `RCL_RET_ACTION_SERVER_TAKE_FAILED` if take failed but no error occurred
+ *         in the middleware, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_take_cancel_request(
+  const rcl_action_server_t * action_server,
+  void * ros_cancel_request);
+
+/// Process a cancel request using a rcl_action_server_t.
+/**
+ * This is a non-blocking call.
+ *
+ * This function will transition one or more active goals to the state CANCELING.
+ * The following cancel policy applies based on the goal ID and the timestamp
+ * contained in the cancel request:
+ *
+ * - If the goal ID is zero and timestamp is zero, cancel all goals.
+ * - If the goal ID is zero and timestamp is not zero, cancel all goals accepted
+ *   at or before the timestamp.
+ * - If the goal ID is not zero and timestamp is zero, cancel the goal with the
+ *   given ID regardless of the time it was accepted.
+ * - If the goal ID is not zero and timestamp is not zero, cancel the goal with the
+ *   given ID and all goals accepted at or before the timestamp.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Yes
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] action_server handle to the action server that will process the cancel request
+ * \param[in] cancel_request a C-typed ROS cancel request to process
+ * \param[out] cancel_reponse a preallocated, zero-initialized, C-typed ROS cancel response
+ *   where the response is copied
+ * \return `RCL_RET_OK` if the response was sent successfully, or
+ * \return `RCL_RET_INVALID_ARGUMENT` if any arguments are invalid, or
+ * \return `RCL_RET_ACTION_SERVER_INVALID` if the action server is invalid, or
+ * \return `RCL_RET_BAD_ALLOC` if allocating memory failed, or
+ * \return `RCL_RET_ACTION_SERVER_TAKE_FAILED` if take failed but no error occurred
+ *         in the middleware, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_process_cancel_request(
+  const rcl_action_server_t * action_server,
+  const rcl_action_cancel_request_t * cancel_request,
+  rcl_action_cancel_response_t * cancel_response);
```

Comment:
@jacobperron same as above, if I'm right about these base action types then I think it'd be nice to have it documented.